internal/appsec/waf/waf.go

```diff
@@ -223,73 +262,81 @@ type Context struct {
 // NewContext a new WAF context and increase the number of references to the WAF
 // handle. A nil value is returned when the WAF handle can no longer be used
 // or the WAF context couldn't be created.
-func NewContext(waf *Handle) *Context {
+func NewContext(handle *Handle) *Context {
 	// Increase the WAF handle usage count by RLock'ing the mutex.
 	// It will be RUnlock'ed in the Close method when the Context is released.
-	waf.mu.RLock()
-	if waf.handle == nil {
-		// The WAF handle got free'd by the time we got the lock
-		waf.mu.RUnlock()
+	handle.refCounter.RLock()
+	if handle.handle == nil {
+		// The WAF waf got free'd by the time we got the lock
+		handle.refCounter.RUnlock()
 		return nil
 	}
-	context := C.ddwaf_context_init(waf.handle)
+	context := C.ddwaf_context_init(handle.handle)
 	if context == nil {
```

Comment:
In this case, we don't increase the number of contexts using the handle, so we should RUnlock() as well, shouldn't we?

Reply:
The WAF handler's `refCounter` is increased for the entire WAF context lifetime, so from a call to `NewContext()` to a call to its `Close()`. If we don't, you can end up release the handle before the context is released (eg. today if you call appsec.Stop() during a request handler execution)

I guess you are asking this because we currently have this long-running lock that would prevent RC updates of the rules-data for a pretty long time, especially when under high loads of handler goroutines.

We should indeed move to the lock-less refcounter implemented in the Sqreen WAF bindings (https://github.com/sqreen/go-libsqreen/blob/master/waf/internal/bindings/waf.go#L79-L93) asap. Let's do this in a separate PR before releasing the IP blocking.

Reply:
I'm talking about the case where `ddwaf_context_init()` returns a nil context. In this case, we have called` RLock()` but don't call `RUnlock()` it seems, and the context `Close()` method won't be called either since there's no context

Reply:
Oh yeah - fixed in https://github.com/DataDog/dd-trace-go/pull/1493/commits/29f48858958df8ffe37ba6c1355bef75bfa6d3ad 👏